tests/functional/test_llm_json.py

```diff
@@ -0,0 +1,362 @@
+                        "location": {
+                            "type": "string",
+                            "description": "City and country e.g. Bogotá, Colombia"
+                        }
+                    },
+                    "required": [
+                        "location"
+                    ],
+                    "additionalProperties": False
+                },
+                "strict": True
+            }
+        },
+        {
+            "type": "function",
+            "function": {
+                "name": "get_pollutions",
+                "description": "Get current level of air pollutions for a given location.",
+                "parameters": {
+                    "type": "object",
+                    "properties": {
+                        "location": {
+                            "type": "string",
+                            "description": "City and country e.g. Bogotá, Colombia"
+                        }
+                    },
+                    "required": [
+                        "location"
+                    ],
+                    "additionalProperties": False
+                },
+                "strict": True
+            }
+        }]
+
+        messages = [
+            {"role": "user", "content": "What is the temperature and pollution level in New York?"}
+        ]
+        completion = client.chat.completions.create(
+            model=model_name,
+            messages=messages,
+            tools=tools,
+            tool_choice="auto",
+        )
+
+        print("COMPLETION:",completion)
+        import json
+        
+        tool_args0 = completion.choices[0].message.tool_calls[0].function.arguments
+        tool_args1 = completion.choices[0].message.tool_calls[1].function.arguments
+        
+        # Assert that tool_call is a valid JSON and matches the schema
+        import json
+        from jsonschema import validate, ValidationError
+
+        try:
+            tool_call_json = json.loads(tool_args0)
+            schema = tools[0]["function"]["parameters"]
+            validate(instance=tool_call_json, schema=schema)
+            tool_call_json = json.loads(tool_args1)
+            schema = tools[1]["function"]["parameters"]
+            validate(instance=tool_call_json, schema=schema)
+            assert True, "tool_call is a valid JSON and matches the schema"
+        except json.JSONDecodeError as e:
+            assert False, f"tool_call is not a valid JSON: {e}"
+        except ValidationError as e:
+            assert False, f"tool_call does not match the schema: {e}"
+
+        messages.append(completion.choices[0].message)
+        messages.append({"role": "tool", "tool_call_id": completion.choices[0].message.tool_calls[0].id, "content": "15 degrees Celsius"})
+        messages.append({"role": "tool", "tool_call_id": completion.choices[0].message.tool_calls[1].id, "content": "pm10 28µg/m3"})
+
+        print("Messages after tool call:", messages)
```

Comment:
should we also assert that tools really get used? are part of the response?

Reply:
there is assertion for the response. tool_calls are included in the response when the tool is used.